syntax = "proto3";

package codex.agent;

// 统一的智能体服务接口
service AgentService {
  // 执行一个智能体任务 (如代码重构、搜索验证等)
  // 这是一个服务器流式 RPC，实时返回任务执行过程中的每一个事件。
  rpc RunTask(RunTaskRequest) returns (stream RunTaskResponse);
}

message RunTaskRequest {
  // 任务请求唯一 ID (用于链路追踪)
  string request_id = 1;

  // 会话唯一 ID (UUID)
  // 用于关联和恢复之前的会话状态
  string session_id = 2;

  // 用户指令 (Prompt)
  string prompt = 3;

  // 会话配置 (严格对齐 Codex ConfigToml 内核)
  SessionConfig session_config = 4;

  // 初始文件上下文
  repeated File context_files = 5;

  // 环境变量覆盖 (用于传递敏感 Key)
  map<string, string> env_vars = 6;
  
  // 基础工作目录
  string base_dir = 7;

  // 历史会话数据 (Codex 原生 JSONL 格式)
  // 如果非空，Adapter 会复活该会话并继续执行
  bytes history_rollout = 8;
}

message SessionConfig {
  string model = 1;
  string model_provider = 2;
  ModelProviderInfo provider_info = 3;
  optional string instructions = 4;
  optional string developer_instructions = 5;
  ApprovalPolicy approval_policy = 6;
  SandboxPolicy sandbox_policy = 7;
  string cwd = 8;
  map<string, McpServerDef> mcp_servers = 9;
}

message ModelProviderInfo {
  string name = 1;
  optional string base_url = 2;
  optional string env_key = 3;
  optional string experimental_bearer_token = 4;
  WireApi wire_api = 5;
  map<string, string> http_headers = 6;
  map<string, string> query_params = 7;
  bool requires_openai_auth = 8;
}

message McpServerDef {
  string server_type = 1; 
  string command = 2;
  repeated string args = 3;
  map<string, string> env = 4;
  string url = 5;
}

enum WireApi {
  WIRE_API_CHAT = 0;
  WIRE_API_RESPONSES = 1;
  WIRE_API_RESPONSES_WEBSOCKET = 2;
}

enum ApprovalPolicy {
  APPROVAL_POLICY_UNSPECIFIED = 0;
  ALWAYS = 1;
  NEVER = 2;
  UNLESS_TRUSTED = 3;
}

enum SandboxPolicy {
  SANDBOX_POLICY_UNSPECIFIED = 0;
  WORKSPACE_WRITE = 1;
  READ_ONLY = 2;
  DANGER_FULL_ACCESS = 3;
}

message File {
  string path = 1;
  bytes content = 2;
}

message RunTaskResponse {
  oneof event {
    // 原始 Codex JSONL 事件
    string codex_event_json = 1;

    // 系统日志 (仅包含关键状态变迁)
    string adapter_log = 2;

    // 错误信息
    string error = 3;

    // 任务成功结束后的最新“灵魂”数据
    bytes updated_rollout = 4;
  }
}
